src/controllers/Order.controller.ts

```diff
@@ -0,0 +1,205 @@
+    if (!userId) {
+      res.redirect('/');
+    }
+    const cart = await getUserCart(userId, res);
+    const page = parseInt(req.query.page as string) || 1;
+    const offset = calculateOffset(page);
+    const [cartItems, totalItems] = await cartItemRepository.findAndCount({
+      where: { cart: { id: cart.id } },
+      relations: ['product'],
+      take: PAGE_SIZE,
+      skip: offset,
+    });
+    const totalPages = Math.ceil(totalItems / PAGE_SIZE);
+
+    // Get total price
+    const rawResult = await cartItemRepository
+      .createQueryBuilder('cartItem')
+      .select('count(cartItem.id) as items')
+      .addSelect('sum(cartItem.quantity * product.price) as total')
+      .leftJoin('cartItem.product', 'product')
+      .where('cartItem.cartId = :cartId', { cartId: cart.id })
+      .getRawOne();
+
+    const { total, items } = rawResult;
+
+    res.render('checkout', {
+      paginationItemsLinks: generatePaginationLinks(page, totalPages),
+      total,
+      items,
+      cartItems,
+    });
+  } catch (err) {
+    console.error(err);
+    next(err);
+  }
+};
+
+const processOrder = async (req: Request, res: Response, imagePath: string) => {
+  // @ts-ignore
+  const user = req.session.user;
+  if (!user) {
+    res.redirect('/');
+  }
+  const { name, phone, address, typeOrder, totalAmount } = req.body;
+  const order = orderRepository.create({
+    name,
+    phone,
+    address,
+    typeOrder,
+    totalAmount,
+    image: imagePath,
+    status: 1,
+    user: user,
+  });
+
+  await orderRepository.save(order);
+
+  const cart = await getUserCart(user.id, res);
+
+  const subquery = cartRepository
+    .createQueryBuilder('cart')
+    .innerJoin('cart.user', 'user')
+    .where('cart.userId = :userId')
+    .getSql();
+
+  const cartItems = await cartItemRepository
+    .createQueryBuilder('cartItem')
+    .innerJoinAndSelect('cartItem.product', 'product')
+    .innerJoinAndSelect(
+      `(${subquery})`,
+      'subquery',
+      'subquery.cart_id = cartItem.cartId',
+      { userId: user.id },
+    )
+    .getMany();
+
+  const orderDetails = [];
+  const products = [];
+  cartItems.forEach((cartItem) => {
+    const orderDetail = new OrderDetail();
+    const product = new Product();
+    orderDetail.quantity = cartItem.quantity;
+    orderDetail.product = cartItem.product;
+    orderDetail.price = cartItem.product.price;
+    orderDetail.order = order;
+    orderDetails.push(orderDetail);
+    const count = cartItem.product.quantity - cartItem.quantity;
+    product.id = cartItem.product.id;
+    product.quantity = count;
+    products.push(product);
+  });
+  await orderDetailRepository.save(orderDetails);
+  await Promise.all(
+    products.map((product) => {
+      productRepository.update(
+        { id: product.id },
+        { quantity: product.quantity },
+      );
+    }),
+  );
+  await cartItemRepository.delete({ cart: { id: cart.id } });
+  res.redirect('/order/success');
+};
```

Comment:
- phần logic này e tách ra các function nhỏ để xử lý nha, nhìn dài dòng rối quá khó maintain sau này e nhé
- đang thao tác với DB rất nhiều bảng dữ liệu -> sử dụng transaction để đảm bảo toàn vẹn dữ liệu